File: processor/src/main/java/com/qualityminds/lazyval/processor/internal/codegen/MapstructGenerator.java
package com.qualityminds.lazyval.processor.internal.codegen;

import com.palantir.javapoet.*;
import com.qualityminds.lazyval.collections.NonEmptySet;
import com.qualityminds.lazyval.processor.spi.Generator;
import com.qualityminds.lazyval.processor.spi.GeneratorResult;
import com.qualityminds.lazyval.processor.spi.StockGeneratorIds;
import com.qualityminds.lazyval.processor.spi.ValidatedGeneratorElement;

import javax.lang.model.element.Modifier;
import java.util.Set;
import java.util.stream.Stream;
import static com.qualityminds.lazyval.processor.internal.codegen.JavaPoetExprs.code;

// must only be public for ServiceLoader, but it is not part of the API
@SuppressWarnings("doclint:accessibility,missing")
final public class MapstructGenerator implements Generator {

    private static final String OPTION_GENERATED_PACKAGE = "lazyval.mapstruct.package";

    @Override
    public String generatorId() {
        return StockGeneratorIds.MAPSTRUCT;
    }

    @Override
    public Set<String> requiredClasspath() {
        return Set.of("org.mapstruct.Mapper");
    }

    @Override
    public Set<String> supportedOptions() {
        return Set.of(OPTION_GENERATED_PACKAGE);
    }

    @Override
    public Stream<GeneratorResult> generate(NonEmptySet<ValidatedGeneratorElement> elements, Context context){
        var mapperAnnotationBuilder = AnnotationSpec.builder(ClassName.get("org.mapstruct", "Mapper"))
                .addMember("unmappedTargetPolicy", "$L", "org.mapstruct.ReportingPolicy.ERROR");

        final String mapperPackage = context.generatorPackage(OPTION_GENERATED_PACKAGE, null);

        TypeSpec.Builder typeSpecBuilder = TypeSpec.interfaceBuilder("LazyvalMapper")
                .addAnnotation(GeneratedStamp.forGenerator(MapstructGenerator.class))
                .addAnnotation(mapperAnnotationBuilder.build())
                .addModifiers(Modifier.PUBLIC);

        elements.forEach(validElement -> {
            typeSpecBuilder.addMethod(buildMapToPayload(validElement));
            typeSpecBuilder.addMethod(buildMapType(validElement));
        });

        JavaFile javaFile = JavaFile.builder(mapperPackage, typeSpecBuilder.build())
                .skipJavaLangImports(true)
                .build();
        return Stream.of(new GeneratorResult.Java(
                new GeneratorResult.Metadata(javaFile.packageName(), javaFile.typeSpec().name()),
                javaFile.toString()));
    }

    private static MethodSpec buildMapType(ValidatedGeneratorElement validElement) {
        var parameterName = "value";

        // The method name and the parameter type ask the same question of either payload; only the
        // null guard differs — a primitive parameter can never arrive null.
        var map = MethodSpec.methodBuilder(String.format("map%sTo%s",
                        validElement.payload().identifier(), validElement.name().flatName()))
                .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                .returns(TypeName.get(validElement.element().asType()))
                .addParameter(TypeName.get(validElement.payloadType()), parameterName);

        if (!validElement.isPayloadPrimitive()) {
            map.beginControlFlow("if(%s == null)".formatted(parameterName))
                    .addStatement("return null")
                    .endControlFlow();
        }

        return map.addStatement("return $L", code(validElement.java().create(parameterName))).build();
    }

    private static MethodSpec buildMapToPayload(ValidatedGeneratorElement validElement) {
        var parameterName = "type";

        var mapToPayload = MethodSpec.methodBuilder(String.format("map%sTo%s",
                        validElement.name().flatName(), validElement.payload().identifier()))
                .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
                .returns(TypeName.get(validElement.payloadType()))
                .addParameter(TypeName.get(validElement.element().asType()), parameterName);

        // Same condition as above, different reason: there the payload was the parameter and a
        // primitive one can never arrive null, here it is the return type and a primitive one cannot
        // carry null back out — `return null` from a method returning `int` would not compile.
        if (!validElement.isPayloadPrimitive()) {
            mapToPayload.beginControlFlow("if(%s == null)".formatted(parameterName))
                    .addStatement("return null")
                    .endControlFlow();
        }

        return mapToPayload
                .addStatement("return $L", code(validElement.java().read(parameterName)))
                .build();
    }
}

File: ksp/src/main/kotlin/com/qualityminds/lazyval/ksp/internal/codegen/MapstructGenerator.kt
package com.qualityminds.lazyval.ksp.internal.codegen

import com.palantir.javapoet.*
import com.qualityminds.lazyval.collections.NonEmptySet
import com.qualityminds.lazyval.ksp.internal.codegen.GeneratedStamp.addGeneratedAnnotation
import com.qualityminds.lazyval.ksp.spi.*
import java.util.stream.Stream
import javax.lang.model.element.Modifier

class MapstructGenerator : Generator {

    companion object {
        private const val OPTION_GENERATED_PACKAGE = "lazyval.mapstruct.package"
    }

    override fun generatorId(): String = StockGeneratorIds.MAPSTRUCT

    override fun requiredClasspath(): Set<String> = setOf("org.mapstruct.Mapper")

    override fun supportedOptions(): Set<String> {
        return setOf(OPTION_GENERATED_PACKAGE)
    }

    override fun generate(
        validatedElements: NonEmptySet<ValidatedKspGeneratorElement>,
        context: Generator.Context
    ): Stream<GeneratorResult> {
        // Build Java interface using JavaPoet
        val interfaceBuilder = TypeSpec.interfaceBuilder("LazyvalMapper")
            .addGeneratedAnnotation(MapstructGenerator::class, context)
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(
                AnnotationSpec.builder(ClassName.get("org.mapstruct", "Mapper"))
                    .addMember("unmappedTargetPolicy", $$"$T.ERROR",
                        ClassName.get("org.mapstruct", "ReportingPolicy"))
                    .build()
            )

        validatedElements.forEach { element ->
            interfaceBuilder.addMethod(createJavaMapToPayloadMethod(element))
            interfaceBuilder.addMethod(createJavaMapFromPayloadMethod(element))
        }

        val packageName = context.generatorPackage(OPTION_GENERATED_PACKAGE, null)

        val javaFile = JavaFile.builder(packageName, interfaceBuilder.build())
            .skipJavaLangImports(true)
            .build()

        return Stream.of(GeneratorResult.Java(
            GeneratorResult.Metadata(packageName, "LazyvalMapper"),
            javaFile.toString()))
    }

    private fun createJavaMapToPayloadMethod(element: ValidatedKspGeneratorElement): MethodSpec {
        // payload rather than the declared payload: a value-class payload is compiled away, so
        // Java only ever sees the type it wrapped. flatName on the domain half too, so two nested
        // types of the same simple name cannot collide into one method signature.
        val methodBuilder = MethodSpec.methodBuilder(
            "map${element.name.flatName()}To${element.payload.identifier()}")
            .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
            .returns(element.payload.toJavaPoet())
            .addParameter(element.name.toJavaPoet(), "type")

        val (read, readArgs) = element.java.read("type").javaPoet()
        if (element.isPayloadPrimitive) {
            methodBuilder.addStatement("return $read", *readArgs)
        } else {
            methodBuilder
                .beginControlFlow("if (type == null)")
                .addStatement("return null")
                .endControlFlow()
                .addStatement("return $read", *readArgs)
        }

        return methodBuilder.build()
    }

    private fun createJavaMapFromPayloadMethod(element: ValidatedKspGeneratorElement): MethodSpec {
        // Resolved by the SPI rather than spelled here: `@JvmName` renames the factory, a companion
        // function without `@JvmStatic` is not on the type at all, and a value-class payload has to go
        // through the generated access shim. None of that is this generator's business.
        val (creation, creationArgs) = element.java.create("value").javaPoet()

        val methodBuilder = MethodSpec.methodBuilder(
            "map${element.payload.identifier()}To${element.name.flatName()}")
            .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
            .returns(element.name.toJavaPoet())
            .addParameter(element.payload.toJavaPoet(), "value")

        if (element.isPayloadPrimitive) {
            methodBuilder.addStatement("return $creation", *creationArgs)
        } else {
            methodBuilder
                .beginControlFlow("if (value == null)")
                .addStatement("return null")
                .endControlFlow()
                .addStatement("return $creation", *creationArgs)
        }

        return methodBuilder.build()
    }

    /**
     * Splits one of the SPI's Java expressions into a JavaPoet format string and the arguments its
     * slots stand for, so type names arrive as imports rather than as fully qualified text.
     */
    private fun PayloadExpr.javaPoet(): Pair<String, Array<Any>> {
        val (format, types) = asFormat($$"$T")
        return format to types.map { it.toJavaPoet() as Any }.toTypedArray()
    }
}
